keel-igor/src/main/kotlin/com/netflix/spinnaker/keel/igor/BuildLifecycleMonitor.kt

```diff
@@ -0,0 +1,197 @@
+import kotlinx.coroutines.runBlocking
+import org.slf4j.LoggerFactory
+import org.springframework.beans.factory.annotation.Value
+import org.springframework.boot.context.properties.EnableConfigurationProperties
+import org.springframework.context.ApplicationEventPublisher
+import org.springframework.stereotype.Component
+import retrofit2.HttpException
+
+@Component
+@EnableConfigurationProperties(LifecycleConfig::class)
+class BuildLifecycleMonitor(
+  override val monitorRepository: LifecycleMonitorRepository,
+  override val publisher: ApplicationEventPublisher,
+  override val lifecycleConfig: LifecycleConfig,
+  val objectMapper: ObjectMapper,
+  val artifactMetadataService: ArtifactMetadataService,
+  val front50Service: Front50Service,
+  @Value("\${spinnaker.baseUrl}") private val spinnakerBaseUrl: String
+) : LifecycleMonitor(monitorRepository, publisher, lifecycleConfig) {
+  private val log by lazy { LoggerFactory.getLogger(javaClass) }
+
+  override fun handles(type: LifecycleEventType): Boolean =
+    type == BUILD
+
+  override suspend fun monitor(task: MonitoredTask) {
+    val buildData = parseAndValidate(task) ?: return
+    kotlin.runCatching {
+      artifactMetadataService.getArtifactMetadata(buildData.buildNumber, buildData.commitId)
+    }.onSuccess { metadata ->
+      val buildMetadata = metadata?.buildMetadata
+      if (buildMetadata == null) {
+        log.error("Error fetching status for $task, response was null")
+        handleFailureFetching(task)
+      } else {
+        when (buildMetadata.status) {
+          "BUILDING" -> publishRunningEvent(task)
+          "SUCCESS" -> publishSucceededEvent(task, buildMetadata)
+          "FAILURE" -> publishFailedEvent(task, buildMetadata)
+          "ABORTED" -> publishAbortedEvent(task, buildMetadata)
+          "UNSTABLE" -> publishSucceededEvent(task, buildMetadata)
+          else -> publishUnknownStatusEvent(task, buildMetadata.status)
+        }
+
+        if (buildMetadata.isComplete()) {
+          endMonitoringOf(task)
+        } else {
+          markSuccessFetching(task)
+        }
+      }
+    }
+      .onFailure { exception ->
+        log.error("Error fetching status for $task: ", exception)
+        handleFailureFetching(task)
+      }
+  }
+
+  /**
+   * Parses build data if format is correct, otherwise publish an Unknown event
+   * and end monitoring of this task.
+   */
+  private fun parseAndValidate(task: MonitoredTask): BuildData? =
+    try {
+      parseBuildData(task)
+    } catch (e: IllegalArgumentException) {
+      publishUnknownEvent(task)
+      endMonitoringOf(task)
+      null
+   }
+
+  private fun parseBuildData(task: MonitoredTask): BuildData =
+      objectMapper.convertValue(task.triggeringEvent.data)
+
+
+  private fun BuildMetadata.isComplete(): Boolean =
+    status == "SUCCESS" || status == "FAILED" || status == "ABORTED"
+
+  private fun publishRunningEvent(task: MonitoredTask) {
```

Comment:
now that we have both bake and builds events, is it possible to extract those methods to a single place to make a cleaner code?

Reply:
It wouldn't be any cleaner. youd still have to pass in the message, and the link, and the original event. so you'd save maybe 1 line of code.

and, it would make it more confusing I think.